### PR TITLE
feat(provisioning): Revert revert of #59496

### PR DESCRIPTION
**Pull request by sentaur-athena**
I added a json field called `metadata` to `SentryApp` in  #59496 which caused [this issue](https://sentry.sentry.io/issues/4611490703/?alert_rule_id=14700499&alert_type=issue&notification_uuid=b306bc43-6452-4029-b26d-3263290b8990&project=1&referrer=slack).

From the issue, it seems like `.distinct` doesn't work with json field and in this case we need to add a key to distinct. I marked the line where the fix is happening.

Discussion in channel for more detail: https://sentry.slack.com/archives/CUHS29QJ0/p1699486403310299

### CODE REVIEW

File: tests/sentry/hybrid_cloud/test_app.py [hunk 1]
```diff
@@ -1,26 +1,44 @@
...
         integration_scopes=["prject:read"],
     )
-    sentry_app_query = SentryApp.objects.all()
-    assert sentry_app_query.count() == 1
+    assert first_app.id == second_app.id
+
+
+@django_db_all(transaction=True)
+@all_silo_test(stable=True)
+def test_find_alertable_services():
```
Comment (sentaur-athena) [inline tests/sentry/hybrid_cloud/test_app.py:32]: This is the test calling disctinct

File: tests/sentry/hybrid_cloud/test_app.py [hunk 2]
```diff
@@ -1,26 +1,71 @@
...
         integration_name="Test Integration",
         integration_scopes=["prject:read"],
     )
-    sentry_app_query = SentryApp.objects.all()
-    assert sentry_app_query.count() == 1
+    assert first_app.id == second_app.id
+
+
+@region_silo_test(stable=True)
+class AppServiceTest(TestCase):
```
Comment (sentaur-athena) [inline tests/sentry/hybrid_cloud/test_app.py:40]: This is the test that calls distinct.

File: src/sentry/services/hybrid_cloud/app/impl.py [hunk 3]
```diff
@@ -90,7 +90,7 @@ def find_alertable_services(self, *, organization_id: int) -> List[RpcSentryAppS
             is_alertable=True,
             installations__status=SentryAppInstallationStatus.INSTALLED,
             installations__date_deleted=None,
-        ).distinct():
+        ).distinct("id"):
```
Comment (sentaur-athena) [inline src/sentry/services/hybrid_cloud/app/impl.py:93]: Here's where the issue happened:
```
django.db.utils.ProgrammingError: UndefinedFunction('could not identify an equality operator for type json\nLINE 1: ..."creator_label", "sentry_sentryapp"."popularity", "sentry_se...\n                                                             ^\n')
SQL: SELECT DISTINCT "sentry_sentryapp"."id", "sentry_sentryapp"."date_deleted", "sentry_sentryapp"."scopes", "sentry_sentryapp"."scope_list", "sentry_sentryapp"."application_id", "sentry_sentryapp"."proxy_user_id", "sentry_sentryapp"."owner_id", "sentry_sentryapp"."name", "sentry_sentryapp"."slug", "sentry_sentryapp"."author", "sentry_sentryapp"."status", "sentry_sentryapp"."uuid", "sentry_sentryapp"."redirect_url", "sentry_sentryapp"."webhook_url", "sentry_sentryapp"."is_alertable", "sentry_sentryapp"."verify_install", "sentry_sentryapp"."events", "sentry_sentryapp"."overview", "sentry_sentryapp"."schema", "sentry_sentryapp"."date_added", "sentry_sentryapp"."date_updated", "sentry_sentryapp"."date_published", "sentry_sentryapp"."creator_user_id", "sentry_sentryapp"."creator_label", "sentry_sentryapp"."popularity", "sentry_sentryapp"."metadata" FROM "sentry_sentryapp" INNER JOIN "sentry_sentryappinstallation" ON ("sentry_sentryapp"."id" = "sentry_sentryappinstallation"."sentry_app_id") WHERE ("sentry_sentryapp"."date_deleted" IS NULL AND "sentry_sentryappinstallation"."date_deleted" IS NULL AND "sentry_sentryappinstallation"."organization_id" = %s AND "sentry_sentryappinstallation"."status" = %s AND "sentry_sentryapp"."is_alertable") LIMIT 21
```

Which doesn't occur after specifying "id" as the key to use for distinct.

**Review by sentaur-athena — COMMENTED** [review]
Everything else is like the reverted PR.

File: src/sentry/migrations/0590_add_metadata_to_sentry_app.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                migrations.RunSQL(
+                    """
+                    ALTER TABLE "sentry_sentryapp" ADD COLUMN "metadata" json NOT NULL DEFAULT '{}';
+                    """,
+                    reverse_sql="""
+                        ALTER TABLE "sentry_sentryapp" DROP COLUMN "metadata";
+                    """,
+                    hints={"tables": ["sentry_sentryapp"]},
+                )
+            ],
```
[the remark this answers is not in the document]
Comment (sentaur-athena) [reply src/sentry/migrations/0590_add_metadata_to_sentry_app.py:45]: Same name and number

**Review by markstory — APPROVED** [review]

**Pull request merged**